scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -0,0 +1,394 @@
+        _bc = self.bc
+        N = len(_shape)
+        L = 0
+        for i in range(N):
+            shape_i = _shape[i]
+            diagonals = np.array(
+                [
+                    np.ones(shape_i, dtype=self.dtype),
+                    -2 * np.ones(shape_i, dtype=self.dtype),
+                    np.ones(shape_i, dtype=self.dtype),
+                ]
+            )
+            if _bc == "Neumann":
+                diagonals[1, 0] = -1
+                diagonals[1, -1] = -1
+            elif _bc in ("Periodic", "Dirichlet"):
+                pass
+            else:
+                raise NotImplementedError(
+                    f"Boundary condition {_bc} is not implemented."
+                )
+
+            offsets = np.array([-1, 0, 1])
+            L_i = spdiags(diagonals, offsets, shape_i, shape_i)
+            if _bc == "Periodic":
+                t = dia_array((shape_i, shape_i))
+                t.setdiag([1], k=-shape_i + 1)
+                t.setdiag([1], k=shape_i - 1)
+                L_i += t
+            else:
+                pass
+            for j in range(i):
+                L_i = kron(eye(_shape[j], dtype=self.dtype), L_i)
+            for j in range(i + 1, N):
+                L_i = kron(L_i, eye(_shape[j], dtype=self.dtype))
+            L += L_i
+        return L
+
+    def _matvec(self, x):
+        _shape = self.dims
+        _bc = self.bc
+        assert np.prod(_shape) == x.shape[0]
+        N = len(_shape)
+        X = x.reshape(_shape + (-1,))
+        Y = -2 * N * X
+        for i in range(N):
+            Y += np.roll(X, 1, axis=i)
+            Y += np.roll(X, -1, axis=i)
+            if _bc in ("Neumann", "Dirichlet"):
+                Y[(slice(None),)*i + (0,) + (slice(None),)*(N-i-1)
+                  ] -= np.roll(X, 1, axis=i)[
+                    (slice(None),) * i + (0,) + (slice(None),) * (N-i-1)
+                ]
+                Y[
+                    (slice(None),) * i + (-1,) + (slice(None),) * (N-i-1)
+                ] -= np.roll(X, -1, axis=i)[
+                    (slice(None),) * i + (-1,) + (slice(None),) * (N-i-1)
+                ]
+                if _bc == "Neumann":
+                    Y[
+                        (slice(None),) * i + (0,) + (slice(None),) * (N-i-1)
+                    ] += np.roll(X, 0, axis=i)[
+                        (slice(None),) * i + (0,) + (slice(None),) * (N-i-1)
+                    ]
+                    Y[
+                        (slice(None),) * i
+                        + (-1,)
+                        + (slice(None),) * (N - i - 1)
+                    ] += np.roll(X, 0, axis=i)[
+                        (slice(None),) * i
+                        + (-1,)
+                        + (slice(None),) * (N - i - 1)
+                    ]
+            elif _bc == "Periodic":
+                pass
+            else:
+                raise ValueError("Unknown boundary condition type "
+                                 f"'{_bc}'.")
```

Comment:
I think we can get rid of all the "unknown boundary condition" type things outside of `__init__`, which already checks we have a known value.

If we want to avoid the case of users assigning a wrong value to `obj.bc`, then we should make it a read-only `@property` with a private value.

Reply:
done